src/dotnet-interactive.Tests/StdIoKernelConnectorTests.cs

```diff
@@ -0,0 +1,153 @@
+            proxyKernel.Name.Should().Be("fsharp2");
+        }
+
+        [Fact]
+        public async Task it_can_create_a_proxy_kernel_to_more_than_one_remote_subkernel()
+        {
+            var connector = CreateConnector();
+            using var rootProxyKernel = await connector.CreateRootProxyKernelAsync();
+
+            var result = await rootProxyKernel.SendAsync(new RequestKernelInfo());
+            var kernelInfos = result.Events.OfType<KernelInfoProduced>().Select(e => e.KernelInfo);
+            var csharpKernelInfo = kernelInfos.Should().ContainSingle(i => i.LanguageName == "C#").Which;
+            var fsharpKernelInfo = kernelInfos.Should().ContainSingle(i => i.LanguageName == "F#").Which;
+
+            using var csharpProxyKernel = await connector.CreateProxyKernelAsync(remoteInfo: csharpKernelInfo);
+            using var fsharpProxyKernel = await connector.CreateProxyKernelAsync(remoteInfo: fsharpKernelInfo, localNameOverride: "fsharp2");
+
+            using var _ = new AssertionScope();
+
+            csharpProxyKernel.Name.Should().Be(csharpKernelInfo.LocalName);
+            csharpProxyKernel.KernelInfo.DisplayName.Should().Be(csharpKernelInfo.DisplayName);
+            csharpProxyKernel.KernelInfo.IsProxy.Should().BeTrue();
+            csharpProxyKernel.KernelInfo.IsComposite.Should().BeFalse();
+            csharpProxyKernel.KernelInfo.LanguageName.Should().Be(csharpKernelInfo.LanguageName);
+            csharpProxyKernel.KernelInfo.LanguageVersion.Should().Be(csharpKernelInfo.LanguageVersion);
+            csharpProxyKernel.KernelInfo.SupportedKernelCommands.Should().BeEquivalentTo(csharpKernelInfo.SupportedKernelCommands);
+            csharpProxyKernel.KernelInfo.SupportedDirectives.Should().BeEquivalentTo(csharpKernelInfo.SupportedDirectives);
+
+            fsharpProxyKernel.Name.Should().Be("fsharp2");
+            fsharpProxyKernel.KernelInfo.DisplayName.Should().Be(fsharpKernelInfo.DisplayName);
+            fsharpProxyKernel.KernelInfo.IsProxy.Should().BeTrue();
+            fsharpProxyKernel.KernelInfo.IsComposite.Should().BeFalse();
+            fsharpProxyKernel.KernelInfo.LanguageName.Should().Be(fsharpKernelInfo.LanguageName);
+            fsharpProxyKernel.KernelInfo.LanguageVersion.Should().Be(fsharpKernelInfo.LanguageVersion);
+            fsharpProxyKernel.KernelInfo.SupportedKernelCommands.Should().BeEquivalentTo(fsharpKernelInfo.SupportedKernelCommands);
+            fsharpProxyKernel.KernelInfo.SupportedDirectives.Should().BeEquivalentTo(fsharpKernelInfo.SupportedDirectives);
+        }
+
+        [Fact]
+        public async Task it_throws_if_proxy_to_the_remote_composite_is_not_created_before_creating_a_proxy_to_a_specific_remote_subkernel()
+        {
+            var connector = CreateConnector();
+            var action = async () => await connector.CreateProxyKernelAsync(remoteName: "csharp");
+            await action.Should().ThrowAsync<InvalidOperationException>();
```

Comment:
⁠I tried to implement the API you had proposed yesterday where `StdIoKernelConnector.CreateProxyKernelAsync()` would create both the root proxy (to the remote composite if it has not been created yet) in addition to the specified 'child' proxy (which is returned from the call e.g., proxy to remote csharp kernel).

This is straightforward to implement. However, if the caller only interacts with the above API (without ever explicitly calling `StdIoKernelConnector.CreateRootProxyKernel`) then the caller will never see the root proxy object. And since the process stays alive until the root proxy (and all other child proxies) are alive, this breaks the ref counted dispose logic. We end up with a zombie process because caller only knows to dispose all child proxies, but not the root proxy...

I considered the following alternatives -

1. Make the connector disposable and add a requirement on the caller to dispose it explicitly. This should in turn dispose the root proxy even if the object was never returned to the caller. The connector was also disposable before the current changes - but as we discussed, it can be confusing to have this requirement.
2. Make `StdIoKernelConnector.CreateProxyKernelAsync()` return both the root and child proxies. Perhaps this could return multiple proxies only for the very first call (i.e. in the case where a root proxy was not present before the call). That would mean changing the signature to retrun a collection of proxies (and potentially changing the name to `CreateProxyKernel*s*Async`). However, I think this also makes the API confusing (especially if we decide to preserve the `CreateRootProxyKernelAsync` API).
3. Require that caller should explicitly call `StdIoKernelConnector.CreateRootProxyKernelAsync()` to fetch the root proxy first. Calls to `StdIoKernelConnector.CreateProxyKernelAsync()` would fail (throw) unless `CreateRootProxyKernelAsync()` has been called explicitly first. In cases where there is just a single kernel on the remote side, `CreateRootProxyKernelAsync()` would be the only call that is needed.

I decided to go with 3 for this PR as I think it results in easier to understand API / semantics than the other options. But I could explore #2 above (or any other suggestions not covered above) if you feel strongly.

Thoughts?

Reply:
The root kernel seems like an implementation detail in this case that we shouldn't expose. Another option could be to dispose the composite proxy internally before returning the subkernel the caller actually asked for. 

This might be more attractive because it gives the externally-facing behavior we want and we can potentially improve the implementation later if we determine a way to do this without needing the composite.

Reply:
Good call - thanks! I'll look into that,